dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/RepositoryImpl.java

```diff
@@ -421,19 +422,21 @@ private NonUniqueResultException excNonUniqueResult(QueryInfo queryInfo,
      * @return exception to replace with, if any. Otherwise, the original.
      */
     @Trivial
-    static RuntimeException failure(Exception original) {
+    static RuntimeException failure(Exception original, EntityManager em) {
         final boolean trace = TraceComponent.isAnyTracingEnabled();
         RuntimeException x = null;
         if (original instanceof PersistenceException) {
             for (Throwable cause = original; x == null && cause != null; cause = cause.getCause()) {
                 if (trace && tc.isDebugEnabled())
                     Tr.debug(tc, "checking " + cause.getClass().getName() + " with message " + cause.getMessage());
 
-                // TODO If this ever becomes real code, it should be delegated to the JDBC integration layer
-                // where there is more thorough logic that takes into account configuration and database differences
-                if (cause instanceof SQLRecoverableException
-                    || cause instanceof SQLNonTransientConnectionException
-                    || cause instanceof SQLTransientConnectionException)
+                if (em != null && cause instanceof SQLException) { //attempt to have the JDBC layer determine if this is a connection exception
+                    WSJdbcDataSource ds = (WSJdbcDataSource) em.unwrap(DataSource.class);
+                    if (ds.getDatabaseHelper().isConnectionError((java.sql.SQLException) cause))
+                        x = new DataConnectionException(original);
+                } else if (cause instanceof SQLRecoverableException
+                           || cause instanceof SQLNonTransientConnectionException
+                           || cause instanceof SQLTransientConnectionException)
```

Comment:
This looks good, but need to account for the possibility that isConnectionError returned false and it is one of the other types. Possibly like this,
```suggestion
                if (em != null && cause instanceof SQLException) { //attempt to have the JDBC layer determine if this is a connection exception
                    WSJdbcDataSource ds = (WSJdbcDataSource) em.unwrap(DataSource.class);
                    if (ds.getDatabaseHelper().isConnectionError((java.sql.SQLException) cause))
                        x = new DataConnectionException(original);
                }
                if (x == null)
                    if (cause instanceof SQLRecoverableException
                           || cause instanceof SQLNonTransientConnectionException
                           || cause instanceof SQLTransientConnectionException)
```